Added create task for creating database

GustavoCaso opened the pull request:
@timriley I was playing around with the new file structure and I thought this `task` might come in handy

What do you think?

GustavoCaso committed:
Added create task for creating database

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -32,6 +32,16 @@ namespace :db do
+      <%= config[:camel_cased_app_name] %>::Container.boot :settings
+      settings = <%= config[:camel_cased_app_name] %>::Container["settings"]
+      database_name = settings.database_url.split('/').last
+      `createdb #{Shellwords.escape(database_name)}`
```
For this to be robust, I think we might also want to pull out the username/password (if any) from the database URL and pass those to the `createdb` command too.

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -32,6 +32,16 @@ namespace :db do
+      settings = <%= config[:camel_cased_app_name] %>::Container["settings"]
+      database_name = settings.database_url.split('/').last
+      `createdb #{Shellwords.escape(database_name)}`
+    end
```
And I reckon it might be good to have an `else` branch here where we print a notice about requiring postgres to be installed (and perhaps `exit` with a non-zero exit code too).

timriley reviewed (changes requested):
Thanks @GustavoCaso :) I reckon it'd be good to get this one in here, and we can always port this logic over to whatever new CLI tool we build in the future, too.

Do you think you could also add a matching `db:drop` task too?

GustavoCaso committed:
Added user and password support for creating db and also support for dropping the db

GustavoCaso commented:
@timriley I update the code to support`username` and `password` as well for `drop` the DB.

I was thinking extracting the logic into a separate class, maybe in `dry-web` but since the `CLI` will probably change in the near future I wanted to know your opinion. 💎

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -33,12 +52,22 @@ namespace :db do
-      `createdb #{Shellwords.escape(database_name)}`
+      execute_database_create_command
+    else
+      `echo you need postgres installed in order to create the database`
```
this could just use `puts "you need postgres..."` instead of shelling out and using the `echo` command 😉

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -15,11 +15,30 @@ def db
+end
+
+def execute_database_create_command(settings)
+  database_url = settings.database_url
```
I wonder if you can get the URL by using `db.url` like elsewhere in this file?

GustavoCaso replied:
I think is not possible, at least for the `create` task, since we need to boot the `rom` and that dependency needs already the database to be created, when trying we get `Sequel::DatabaseConnectionError: PG::ConnectionBad: FATAL`

I could use it for the `drop`

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -15,11 +15,30 @@ def db
+
+def execute_database_create_command(settings)
+  database_url = settings.database_url
+  _, _, _, db_name = database_url.split('/')
```
I think it might be safer if we could use `URI.parse` for this instead of trying to split on a character?

That'd also let us extract the username and password cleanly, too.

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -15,11 +15,30 @@ def db
+def execute_database_create_command(settings)
+  database_url = settings.database_url
+  _, _, _, db_name = database_url.split('/')
+  if user_and_net.include?('@')
```
I can't see where `user_and_net` is assigned? Did you leave something out, perhaps? Anyway, perhaps it might be worth trying the URI.parse approach I mentioned above.

GustavoCaso replied:
Yes, you are right I definitely left out that, part. I would change it anyway to use `URI`

timriley commented on lib/dry/web/roda/templates/Rakefile.tt:
```diff
@@ -15,11 +15,30 @@ def db
+  <%= config[:camel_cased_app_name] %>::Container["settings"]
+end
+
+def execute_database_create_command(settings)
```
Since we're only using this in one place, I don't mind this being right inside the db:create task rather than in a method like this.

timriley reviewed (commented):
Few more notes for you, @GustavoCaso :)

GustavoCaso committed:
Apply feedback, use URI to parse url

solnic commented:
IIRC hanami-model has CLI for this, we could just adopt it in rom-sql

GustavoCaso commented:
Any update on this one @timriley @solnic?

timriley committed:
Simplify naming of db URL var

timriley committed:
Specify all Postgres connection info with vars when creating/dropping DB

timriley committed:
Tweak wording across rake tasks

timriley committed:
Update db:structure:dump task to use postgres env vars

timriley committed:
Pass postgres env vars to commands in cleaner way

Using `system` with a hash of env vars hopefully means that they can’t be found by inspecting `ps` output while the command is running.

timriley commented:
@GustavoCaso Thanks for your work on this! I've just tidied things up a little more (e.g. I discovered a way to pass more of the postgres connection info via env vars, which lets us handle this consistently across all postgres commands). I'll pull it in now and get it out in a release today.
